src/sessions/middleware.rs

```diff
@@ -0,0 +1,255 @@
+/// sessions would still check the expiry on the contained session
+/// before using it
+///
+/// ### If anything goes wrong with the above process
+///
+/// If there are any failures in the above session retrieval process,
+/// a new empty session is generated for the request, which proceeds
+/// through the application as normal.
+///
+/// ## Stale/expired session cleanup
+///
+/// Any session store other than the cookie store will accumulate
+/// stale sessions.  Although the tide session middleware ensures that
+/// they will not be used as valid sessions, For most session stores,
+/// it is the tide application's responsibility to call cleanup on the
+/// session store if it requires it (not all session stores do -- for
+/// example, redis and memcached support expiry natively).
+
+pub struct SessionMiddleware<Store> {
+    store: Store,
+    cookie_path: String,
+    cookie_name: String,
+    session_ttl: Option<Duration>,
+    key: Key,
+}
+
+impl<Store: SessionStore> std::fmt::Debug for SessionMiddleware<Store> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SessionMiddleware")
+            .field("store", &self.store)
+            .field("cookie_path", &self.cookie_path)
+            .field("cookie_name", &self.cookie_name)
+            .field("session_ttl", &self.session_ttl)
+            .field("key", &"..")
+            .finish()
+    }
+}
+
+impl SessionMiddleware<MemoryStore> {
+    /// Builds a new SessionMiddleware with an ephemeral in-memory
+    /// store. See [`MemoryStore`](crate::session::MemoryStore) for
+    /// important caveats and limitations of this store.
+    /// DO NOT USE THIS STORE IN PRODUCTION
+    pub fn memory_store(secret: &[u8]) -> Self {
+        Self::new(MemoryStore::new(), secret)
+    }
+}
+
+impl SessionMiddleware<CookieStore> {
+    /// Builds a new SessionMiddleware with a cookie store. See
+    /// [`CookieStore`](crate::session::CookieStore) for important
+    /// caveats and limitations of this store.
+    pub fn cookie_store(secret: &[u8]) -> Self {
+        Self::new(CookieStore::new(), secret)
+    }
+}
+
+#[async_trait]
+impl<Store, State> Middleware<State> for SessionMiddleware<Store>
+where
+    Store: SessionStore,
+    State: Send + Sync + 'static,
+{
+    async fn handle(&self, mut request: Request<State>, next: Next<'_, State>) -> crate::Result {
+        let cookie_value = request
+            .cookie(&self.cookie_name)
+            .and_then(|cookie| self.verify_signature(cookie.value()).ok());
+        let mut session = self.load_or_create(cookie_value).await;
+
+        if let Some(ttl) = self.session_ttl {
+            session.expire_in(ttl);
+        }
+
+        let secure_cookie = request.url().scheme() == "https";
+        request.set_ext(session.clone());
+
+        let mut response = next.run(request).await;
+
+        if let Some(cookie_value) = self.store.store_session(session).await {
```

Comment:
This is a bug (unless `Session` is in an `Arc`).

You are storing a pre-cloned session, anything set will be lost.

Above should be `request.set_ext(session);` and this should actually be a `Session` from `request.take_ext()` (does not yet exist).

Reply:
this is not true. look at the implementation of Session. it uses interior mutability around the data. the only data that might be lost is the expiry, and it is not currently  part of the intended api for applications to modify the expiry directly

Reply:
Ugh hmm, that pattern is quite confusing, I'll have to give it a look again soon.

Reply:
Would really like to remove the interior mutability, but without being able to retrieve the Ext, there has to be an inner Arc<RwLock<_>> in which there's no need for a mutable outer reference. Marking as resolved, since this issue has been turned into [an http-types pr for discussion](https://github.com/http-rs/http-types/pull/201)